### PR TITLE
Core/Spells: fix wrong distance calculations in AoE spells [Needs testing]

### PR DESCRIPTION
**Pull request by chaodhib**
This PR attempts to fix an issue encountered in several places including here: https://github.com/TrinityCore/TrinityCore/pull/15008 regarding AoE spells. The AoE calculations are broken. The size of the aura's dynamic object and the size of the target are taken into account which shouldn't happen.

Changes:
- the calculation doesn't take into account the size of the dynamic object (the object at the center of the aoe) anymore
- the calculation doesn't take into account the size of the player anymore
- the distance calculation is done now in 2D.

### CODE REVIEW

File: src/server/game/Entities/Object/Object.cpp
Comment (jackpoz) [inline src/server/game/Entities/Object/Object.cpp]: isn't this the same as https://github.com/TrinityCore/TrinityCore/pull/16290/files#diff-29afb15ead2a8204cac07d94f074f260R1114 with the last 2 arguments are optional ?

**Comment by jackpoz** [comment]
does changing the dist calc to 2D mean a very high Z diff will still make the AoE spell affect the target ? is that how it's supposed to work ?

File: src/server/game/Entities/Object/Object.cpp
Comment (chaodhib) [inline src/server/game/Entities/Object/Object.cpp]: WorldObject::_IsWithinDist(WorldObject const* obj, float dist2compare, bool is3D) const is overriden in GameObject. I didn't want to rewrite that method because I'm not sure what it does exactly/ how it does it.

So I created a separate method in order to not affect the rest of the code. Maybe not the best approach..

**Comment by Treeston** [comment]
You tracked down that bug? Nice :+1:

**Comment by chaodhib** [comment]
@jackpoz I'm beginning to have doubts. I'm not sure

**Comment by jackpoz** [comment]
I have no idea how your code behaves with high Z diff and how was the blizzlike behavior but it's worth to investigate

**Comment by chaodhib** [comment]
it's not the main concern of this PR. I could drop it. But yeah, That would be also interesting to know.

**Comment by Aokromes** [comment]
As for frost nova it can hit flying player few around 6-8 yards over floor.

File: src/server/game/Entities/Object/Object.cpp
Comment (jackpoz) [inline src/server/game/Entities/Object/Object.cpp]: little typo here "// sie"

**Comment by Treeston** [comment]
@Aokromes can you check if it's a sphere or just a xy ring plus z range?

E.g. is the height at which you can hit less or or the same at the edge vs near the center?

**Comment by Aokromes** [comment]
Can't anymore, ptr is down, and it will very likely will be down for months.

**Comment by Treeston** [comment]
Could someone with a retail account check this then, please? Only thing left before this can get merged.

If we can't, then @chaodhib just assume it's a 2d check plus z diff check (not sphere, just check if abs(targetz-ownz) <= aoe radius in addition existing 2d check) and we merge that.

**Comment by chaodhib** [comment]
I rather not rush on this PR. I know it's affecting a lot of encounters but I would like to have more data before we implement a new implementation.

**Comment by Treeston** [comment]
The difference between xyz sphere and xy sphere + z range is miniscule in 99% of cases.

**Comment by jackpoz** [comment]
is the 2D/3D the only unclear thing ? it could be restored to be 3D so the rest can be merged

**Comment by Treeston** [comment]
Just make it a cylinder imo, that's what it "feels like" on retail (super inaccurate, I know, honestly it makes no practical difference).

**Comment by chaodhib** [comment]
I agree with the cylinder part, except it wont work great in a non flat area.

Currently, we have this: (which is not precise but ok)
gtaoe cylindar and sphere non flat2b
Using a cylinder, we'll have this: (which is bad)
gtaoe cylindar and sphere non flat2a
A better solution would be this:
gtaoe cylindar and sphere non flat2c
which is a "cylinder above the ground". So we'll need to use MMAPs for this.

Or we could use a double cylinder (one above and one under the horizontal plane). Would be less precise but would work too. The simpler calculations would be good for the speed. However, we need to ask ourselves if this approximation is good enough. (the answer to that question depends on the maximum slope a player or NPC can go upward. if that value is low enough, we'll never encounter the situation in the picture below and therefore we'll be fine)
gtaoe cylindar and sphere non flat2d

My point is, the 2D/3D subject is harder than it looks. I prefer to drop it from this PR and focus on the main issue.

Regarding the main issue: Quick reminder; AoEs can be divided into 2 categories: GTAoE (ground targeted AoE, example: Blizzard, Shadowfury,...) and PBAoE (proximity based AoE. example: Frost Nova, the hunter trap I suppose, etc...). Can we say that:
- in case of a GTAoE, nor the target's radius nor the caster's radius are to be taken into account in the distance calculation?
- in case of a PBAoE, only the caster's radius are to be taken into account?

In both case, do we have proof?

PS: there seems to be an open issue linked to this PR: https://github.com/TrinityCore/TrinityCore/issues/3923

**Comment by jackpoz** [comment]
The cylinder in 2nd screenshot should have same height as the sphere radius, that way the diff in volume between sphere and cylinder is very little

**Comment by Treeston** [comment]
Target hitbox has to be taken into account btw. Otherwise, y'know, dragon bosses.

Came to mind during Sapphiron today :D

**Comment by Keader** [comment]
Testing this PR in Halion...
Some spells still dont hit he in right distance.
Divine storm (8 yards - spellid:  53385) and Whirlwind (8 yards - spellid:  50622).

Using .dist on Boss:
The distance is: (3D) 6.415426 (2D) 6.415344 (Exact 3D) 25.915426 (Exact 2D) 25.915344 yards. 
_Works, boss get damage._
But...
The distance is: (3D) 6.761860 (2D) 6.761780 (Exact 3D) 26.261860 (Exact 2D) 26.261780 yards. 
_Dont work, boss dont get any damage._

**Comment by Keader** [comment]
Any news about https://github.com/TrinityCore/TrinityCore/pull/16290#issuecomment-196499053 ?

**Comment by sirikfoll** [comment]
@chaodhib , must be due to the 3d->2d thing, that this issue happens:
go to `.go xyz -553 2200 51 631` and use death knight's spell Death and Decay, the spell will hit npcs that are on the upper floor, this seems to occur for any spells with `SPELL_EFFECT_PERSISTENT_AREA_AURA`.

**Comment by chaodhib** [comment]
I'm working on this but I've nothing concrete yet. I'm doing sniff analysis to determine the formulas used in retail to calculate the ranges. I got tons of data but nothing makes much sense to me so far. I'll post the data soon.

**Comment by Aokromes** [comment]
Maybe the way to go is ingame inspection while 2 characters does things.

**Comment by chaodhib** [comment]
@Aokromes What do you mean? You mean doing tests on retail with 2 players while sniffing? If so, I agree.

I managed to get the right formulas for the auto attack attack, melee spells but for aoes, I can't use the same trick since the client has no aoe range validation. 

Instead what I did was wrote a software that analyze sniffs and look for times when an unit is hit by a spell at a distance higher than the range of the spell. I managed to have a working version and got some data but I've found any way to interpret these data. I'll dump as much data as I can when I'll find time, probably not before the end of this month.

On the other hand, taking the approach of doing tests on a live retail server would be probably faster and better..

**Comment by Aokromes** [comment]
I mean with 3, 2 doing test and a 3rd looking the tests.

**Comment by chaodhib** [comment]
When doing the test, we need the exact positions (of the players and the spell origin), which can only be obtained by parsed sniffs. Therefore, only 2 players with at least one of them sniffing is enough I think

**Comment by tkrokli** [comment]
@chaodhib : good job working on this. Do you have any opinions to share in issue #15151 ? :-)

**Comment by chaodhib** [comment]
AoE spells are more complicated than I thought. I think that the formula that determines if a unit is in range of an AoE actually depends on the ImplicitTarget flags of the aoe spell.

My current understanding is that:
- for GTAoEs, the formula is `range = spell radius + caster combat reach + target combat reach`
- for PBAoE, the formula is `range = spell radius`
- for AoEs created by dynamic objects, the formula is : `range = spell radius`. 

To make things smoother, this PR will only fix dynObj based AoEs.

**Comment by Treeston** [comment]
This would prevent mages from hitting, say, dragon bosses with Arcane Explosion (it's PBAoE, which wouldn't account for hitbox). I agree for GTAoE and dynobj AoEs, but not for PBAoE.

**Comment by chaodhib** [comment]
@Treeston My bad. I mistakenly reversed the 2. I fixed it now.

**Comment by Treeston** [comment]
Hm, that means that DnD would only hit a dragon boss if you cast it on its center, not its nose? Seems weird, but I can't remember how it worked.

Should be easily to verify on retail though.

**Comment by chaodhib** [comment]
DnD uses a dynObj

**Comment by Treeston** [comment]
Right. What's an example of GTAoE? Did you test target hitbox for dynobj AoE separately?

**Comment by chaodhib** [comment]
Actually strike that, even if DnD uses a dynObj, it's the same formula as GTAoEs, so yeah, if i'm right, you would need to cast the spell pretty close to the center of the unit.

The reason I think that dynObj AoEs and GT AoEs share the same formula is spells like Flamestrike. This spell has 2 spell effect. One which uses TARGET_DEST_DYNOBJ_ENEMY (dynObj formula) for the dot component. And the other which uses TARGET_UNIT_DEST_AREA_ENEMY (GTAoE formula) for the direct damage component. That's not a full proof but it would make sense that the 2 spell effects have the same effective radius.

**Comment by Treeston** [comment]
Yeah, but I'm wondering if they might disregard caster (dynobj) but take into account target (potentially large dragon boss).

Like I said, should be easy to test on retail - just walk up to Sapphiron or something and drop a DnD so it only overlaps the edge of his hitbox.

**Comment by chaodhib** [comment]
@Treeston Yes. It's possible that it takes into account the target combat reach as well, though my current understand is that it does not.

Would be cool if someone with a retail account could test that (and record it if possible).

**Comment by Treeston** [comment]
@Aokromes ?

**Comment by Aokromes** [comment]
https://onedrive.live.com/redir.aspx?cid=83725776ab2e4745&resid=83725776AB2E4745!1874&parId=83725776AB2E4745!143&authkey=!AHJWo0hk4QSBwLw&ithint=video%2cmp4

**Comment by Treeston** [comment]
So it accounts for target, but doesn't account for caster. Makes sense.

@chaodhib can we get this finished with that info?

**Comment by chaodhib** [comment]
So, if I understan correctly, these are the formulas:

- for PBAoEs, the formula is `range = spell radius + caster combat reach + target combat reach`
- for GTAoEs, the formula is `range = spell radius + target combat reach`
- for AoEs created by dynamic objects, the formula is : `range = spell radius + target combat reach`

I made the change for dynObj based AoEs. I'll work on the rest later.

**Comment by Treeston** [comment]
In more simple terms:
* Is it PBAoE? Then the AoE applies from edge of our hitbox.
* Is it non-PB AoE effect? Then the AoE applies as a circle from the destination point.

In both cases, any target whose hitbox overlaps the AoE will be hit.

**Comment by Treeston** [comment]
@chaodhib is this ready for merge yet?

**Comment by chaodhib** [comment]
@Treeston No, I need time to review this and I need to analyze more sniffs of woltk to confirm what was discussed here. Sadly, I won't have much available time in the following weeks.

**Comment by Keader** [comment]
@chaodhib This PR needs update :)

**Comment by chaodhib** [comment]
Since the combat radius of players is so small, I don't think the impact of this issue is big. Also, it's very hard to prove the correctness of the new formulas, using sniff data. And believe me, I tried.

**Comment by Treeston** [comment]
Can we just make the logic assume what I stated in [this comment](https://github.com/TrinityCore/TrinityCore/pull/16290#issuecomment-237676040) (which is the most accurate guess we have), @chaodhib?

Then review & merge.

**Comment by chaodhib** [comment]
Ok. Will do that.

**Comment by Treeston** [comment]
Actually, doesn't that just mean that we should give dynobjects a hitbox radius of zero?

**Comment by Aokromes** [comment]
Conflicting files
src/server/game/Entities/Object/Object.h
src/server/game/Grids/Notifiers/GridNotifiers.h
src/server/game/Spells/Auras/SpellAuras.cpp

**Comment by chaodhib** [comment]
Alright, I'm more confident about the formulas that we came up with now (though still not 100% sure). I'm gonna start to work on the implementation.

**Comment by Treeston** [comment]
> Actually, doesn't that just mean that we should give dynobjects a hitbox radius of zero?
Either way, awesome :+1:

**Comment by chaodhib** [comment]
I think dynObj do not have a size on their own. Their size is the spell radius. Another way to see this is to say that their hitbox radius is zero, yeah. Either way, it's the same thing.

**Comment by chaodhib** [comment]
Done. Ready for extensive testing and review.

**Comment by Treeston** [comment]
:heart:

File: src/server/game/Spells/Auras/SpellAuras.cpp
Comment (Keader) [inline src/server/game/Spells/Auras/SpellAuras.cpp]: we not use nullptr instead NULL?

Comment (Treeston) [inline src/server/game/Spells/Auras/SpellAuras.cpp]: :+1: `nullptr` please

Comment (chaodhib) [inline src/server/game/Spells/Auras/SpellAuras.cpp]: I'm not sure. I thought it was the opposite? In existing code, it can see both.

Comment (Aokromes) [inline src/server/game/Spells/Auras/SpellAuras.cpp]: NULLs are moving into nullptr

Comment (chaodhib) [inline src/server/game/Spells/Auras/SpellAuras.cpp]: ok

**Comment by chaodhib** [comment]
To synthesize, here are the changes:

**BEFORE:**
- for PBAoEs, `effective range = spell radius + target combat reach`
- for GTAoEs, `effective range = spell radius + target combat reach`
- for AoEs created by UnitAura, `effective range = spell radius + target combat reach` + caster combat reach
- for AoEs created by DynObjAura, `effective range = spell radius + target combat reach` + DEFAULT_WORLD_OBJECT_SIZE

**AFTER:**
- for PBAoEs, `effective range = spell radius + caster combat reach + target combat reach`
- for GTAoEs, `effective range = spell radius + target combat reach`
- for AoEs created by Auras, `effective range = spell radius + target combat reach`

**Comment by Aokromes** [comment]
@chaodhib  irc :P

**Comment by Treeston** [comment]
Alright, as per @Aokromes test results on IRC, these are the hit formulas:

```range = radius + target hitbox radius
if PBAoE:
    range += caster hitbox radius
target is affected iff ((distance2d(origin, target) <= range) AND (abs(origin.z-target.z) <= range))```

**Comment by chaodhib** [comment]
Could I get access to the data?

**Comment by chaodhib** [comment]
Done. Awaiting for test & review.

**Comment by Aokromes** [comment]
@chaodhib this can be merged? if yes and no one tells something to stop the merge i will merge it tomorrow.

**Comment by chaodhib** [comment]
@Aokromes Apparently, there is no negative feedback. So yeah, in my opinion, it's ready to be merged.

**Comment by ccrs** [comment]
correct codestyle first

**Comment by chaodhib** [comment]
@ccrs could you be more specific please?

**Comment by tkrokli** [comment]
I read through all of the changes in the files, but could not find any obvious codestyle issues, aside from minor column indent variations in some of the `#define`s in a couple of the `.h` files (but that's not important). I also noticed that `IsWithinDoubleVerticalCylindar` is a tiny typo, but that doesn't matter either because all 5 of them have the same name. So yes, need to have any codestyle issues pointed out.

File: src/server/game/Entities/Object/Position.h
Comment (Treeston) [inline src/server/game/Entities/Object/Position.h]: This is unused and should be removed.

File: src/server/game/Entities/Object/Position.h
Comment (Treeston) [inline src/server/game/Entities/Object/Position.h]: Just `dist2d < radius && abs(dz) < height` summarizes it well enough

File: src/server/game/Entities/Object/Position.h
Comment (Treeston) [inline src/server/game/Entities/Object/Position.h]: Is this still used anywhere? If not, remove.

File: src/server/game/Entities/Object/Position.h
Comment (Treeston) [inline src/server/game/Entities/Object/Position.h]: Refactor `Cylindar` -> `Cylinder`

**Review by Treeston — APPROVED** [review]
Final CS adjustments, then merge.

File: src/server/game/Spells/Spell.cpp
Comment (ccrs) [inline src/server/game/Spells/Spell.cpp]: here

File: src/server/game/Spells/Spell.cpp
Comment (ccrs) [inline src/server/game/Spells/Spell.cpp]: maybe a comment is better than using a local variable to explain what the check does, I dont know if there is a convention about this

File: src/server/game/Grids/Notifiers/GridNotifiers.h
Comment (ccrs) [inline src/server/game/Grids/Notifiers/GridNotifiers.h]: here

File: src/server/game/Grids/Notifiers/GridNotifiers.h
Comment (ccrs) [inline src/server/game/Grids/Notifiers/GridNotifiers.h]: here

File: src/server/game/Grids/Notifiers/GridNotifiers.h
Comment (ccrs) [inline src/server/game/Grids/Notifiers/GridNotifiers.h]: here

File: src/server/game/Entities/Object/Object.cpp
Comment (ccrs) [inline src/server/game/Entities/Object/Object.cpp]: better use other name for the local variables (a & b), everything would be more readable

Comment (ariel-) [inline src/server/game/Entities/Object/Object.cpp]: Convention is lowerCamelCase for local variable, const after type not before.

File: src/server/game/Entities/Object/Position.h
Comment (chaodhib) [inline src/server/game/Entities/Object/Position.h]: it's used somewhere

File: src/server/game/Entities/Object/Position.h
Comment (chaodhib) [inline src/server/game/Entities/Object/Position.h]: ok

File: src/server/game/Entities/Object/Position.h
Comment (chaodhib) [inline src/server/game/Entities/Object/Position.h]: ok

File: src/server/game/Entities/Object/Position.h
Comment (chaodhib) [inline src/server/game/Entities/Object/Position.h]: ok

File: src/server/game/Spells/Spell.cpp
Comment (chaodhib) [inline src/server/game/Spells/Spell.cpp]: what is wrong here?

File: src/server/game/Spells/Spell.cpp
Comment (chaodhib) [inline src/server/game/Spells/Spell.cpp]: i disagree

File: src/server/game/Grids/Notifiers/GridNotifiers.h
Comment (chaodhib) [inline src/server/game/Grids/Notifiers/GridNotifiers.h]: you mean the statement on a different line then the if, right?

File: src/server/game/Entities/Object/Object.cpp
Comment (chaodhib) [inline src/server/game/Entities/Object/Object.cpp]: ok

File: src/server/game/Spells/Spell.cpp
Comment (Treeston) [inline src/server/game/Spells/Spell.cpp]: `if (`

File: src/server/game/Grids/Notifiers/GridNotifiers.h
Comment (Treeston) [inline src/server/game/Grids/Notifiers/GridNotifiers.h]: yes

File: src/server/game/Entities/Object/Object.cpp
Comment (chaodhib) [inline src/server/game/Entities/Object/Object.cpp]: any idea of names I could use for a and b?

**Comment by chaodhib** [comment]
I think that's all

File: src/server/game/Spells/Spell.cpp
Comment (chaodhib) [inline src/server/game/Spells/Spell.cpp]: ty

File: src/server/game/Spells/Spell.cpp
Comment (ccrs) [inline src/server/game/Spells/Spell.cpp]: Given the past dramas I understand the... answer I guess 😄
But no worries, it was more of a headsup to open a debate about it since its one of the few times I've seen this type of codestyle in the core and actually anywhere else (using codestyle to show meaning instead of comments)

Comment (DDuarte) [inline src/server/game/Spells/Spell.cpp]: Avoid comments but use them when you *need* them.

https://blog.codinghorror.com/coding-without-comments/

> While comments are neither inherently good or bad, they are frequently used as a crutch. You should always write your code as if comments didn't exist. This forces you to write your code in the simplest, plainest, most self-documenting way you can humanly come up with.

**Pull request merged**
